### PR TITLE
Make rendering encoding configurable

### PR DESCRIPTION
**Pull request by dataflake**
This replaces branch `encoding-as-optional-argument` and is based on current master, I broke that branch in a botched rebase operation.

This will start fixing zopefoundation/Zope#271

From Malthe's original PR #23:
This adds an optional keyword argument encoding to the publicly exposed rendering functions and classes.

See #20 for previous discussion.
Closes #20.
Closes #22.
Closes #23.

### CODE REVIEW

**Comment by icemac** [comment]
Hm, the tests are only green on Python 2 and there are some Flake 8 errors. @dataflake Are you planning to look into them?

**Comment by dataflake** [comment]
This is work in progress and you know it's done when I ask for a review.

**Comment by dataflake** [comment]
I just realized I had already sent a review request, sorry, that was premature.

File: src/DocumentTemplate/_DocumentTemplate.py [hunk 1]
```diff
@@ -103,83 +103,80 @@
...
-                if not isinstance(t, basestring):
+                if not isinstance(t, (six.string_types, six.binary_type)):
                     # This might be a TaintedString object
                     untaintmethod = getattr(t, '__untaint__', None)
                     if untaintmethod is not None:
                         # Quote it
                         t = untaintmethod()
                         skip_html_quote = 1
 
-                if not isinstance(t, basestring):
```
Comment (icemac) [inline src/DocumentTemplate/_DocumentTemplate.py]: How did this work with all the `unicode` and `basestring` names?

Comment (icemac) [reply src/DocumentTemplate/_DocumentTemplate.py]: Ah, I see, we were defining them at the beginning of the module.

Comment (dataflake) [reply src/DocumentTemplate/_DocumentTemplate.py]: There's many places where code wants to decide "is this a string" as opposed to, say, a callable. The work that's been done to make all that Python 3 compatible was really haphazard, inconsistent and incomplete. I've spent hours stepping through the debugger on this, and I am dure there will be more edge cases found in the future.

**Pull request merged**